README.md
# BytesIOEx

![PyPI](https://img.shields.io/pypi/v/bytesioex)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/bytesioex)
[![Code Style: Black][black]](https://github.com/psf/black)
[![security: bandit][bandit]](https://github.com/PyCQA/bandit)

BytesIOEx is a simple wrapper over Python's `io.BytesIO` which provides
additional methods for reading and writing C data types like **int8**,
**uint8**, **bool** and so on. The `read_*` methods are used for reading a
particular type from the stream and the `write_*` methods are used for writing
Python's basic data types **int**, **bool** and **float** to the stream. Both
these types of methods advance the steam position by the size of the data type.
The type conversion is handled by the `struct` module. `Struct` classes are
used to maximize the performance. **Native byteorder is used.**

## ⏬ Installation

BytesIOEx requires Python 3.6+, it may run on older version as well but I have
not tested it.

```
pip install --upgrade bytesioex
```

## 📜 Documentation

Docs are available on [ReadTheDocs][docs].

## 🚀 Roadmap

- [ ] Ensure full coverage.
- [ ] Add `wchar` (UTF-16 character) support.
- [ ] Support for different byte orders.

## ❕ Motivation

C# `BinaryReader` and `BinaryWriter`.

## 📧 Contact

E-mail: demberto@protonmail.com

## © License

BytesIOEx is distributed under the [MIT License][license].

<!-- BADGES -->
[bandit]: https://img.shields.io/badge/security-bandit-yellow.svg
[black]: https://img.shields.io/badge/code%20style-black-black

<!-- PROJECT LINKS -->
[docs]: https://bytesioex.rtfd.io/en/latest/
[license]: https://github.com/demberto/BytesIOEx/blob/master/LICENSE


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.2] - 2022-05-29

### Added

- Strict type checking for code and tests with `mypy`.

### Fixed

- Moved `py.typed` to correct location ensuring compatibility with mypy.
- CI.

## [0.1.1] - 2022-03-20

Better code quality enforcements and a leaner project structure. BytesIOEx
boasts of a 98% coverage now!

### Added

- PEP-561 compliance (`py.typed`).
- More pre-commit hooks for maintaining code quality.
- Adopted semantic versioning.
- `tbump` for version handling.
- Flake8, Bandit and Pylint.
- Follow Google-style Python coding guidelines.

### Changed

- `requirements` file now has a constraints file.
- All old tasks replaced with `tox-gh-actions` in CI.
- Beautified `README.md` (and `docs/index.md`).

### Removed

- `conftest.py`. `pytest` config is now in `pyproject.toml`.
- `TODO.md`; its contents have been moved to `README.md` > Roadmap.
- `docs/reference.md`; contents moved to `docs/index.md`.

## [0.1.0]

### Added

- Automated build and publish on commit
- [Docs](https://bytesioex.rtfd.io)

### Changed

- Smaller function names.

### Fixed

- Import errors.
- Code cleanup.

<!-- VERSION LINKS -->

[0.1.2]: https://github.com/demberto/bytesioex/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/demberto/bytesioex/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/demberto/bytesioex/releases/tag/v0.0.1


test.py
import unittest

from bytesioex.bytesioex import BytesIOEx

class TestBytesIOEx(unittest.TestCase):
    def test_read_bool(self):
        stream = BytesIOEx(b'\x00')
        self.assertEquals(stream.read_bool(), False)
    
    def test_write_bool(self):
        stream = BytesIOEx()
        stream.write_bool(True)
        stream.seek(0)
        self.assertEquals(stream.read_bool(), True)
    
    # TODO: More tests

tests/test_bytesioex.py
"""Tests `BytesIOEx` class."""

import math
import random
from typing import Callable

import pytest

from bytesioex import (
    Bool,
    Byte,
    BytesIOEx,
    Double,
    Float,
    Int,
    Long,
    SByte,
    Short,
    UInt,
    ULong,
    UShort,
)

INT8_MIN = -(2**7)
INT8_MAX = (2**7) - 1
UINT8_MAX = (2**8) - 1

INT16_MIN = -(2**15)
INT16_MAX = (2**15) - 1
UINT16_MAX = (2**16) - 1

INT32_MIN = -(2**31)
INT32_MAX = (2**31) - 1
UINT32_MAX = (2**32) - 1

INT64_MIN = -(2**63)
INT64_MAX = (2**63) - 1
UINT64_MAX = (2**64) - 1

UINT8_MIN = UINT16_MIN = UINT32_MIN = UINT64_MIN = 0

CONVERTER = {
    "B": Byte,
    "b": SByte,
    "H": UShort,
    "h": Short,
    "I": UInt,
    "i": Int,
    "f": Float,
    "d": Double,
    "q": Long,
    "Q": ULong,
}


@pytest.mark.parametrize(
    "min_, max_, fmt",
    [
        (UINT8_MIN, UINT8_MAX, "B"),
        (INT8_MIN, INT8_MAX, "b"),
        (UINT16_MIN, UINT16_MAX, "H"),
        (INT16_MIN, INT16_MAX, "h"),
        (UINT32_MIN, UINT32_MAX, "I"),
        (INT32_MIN, INT32_MAX, "i"),
        (UINT64_MIN, UINT64_MAX, "Q"),
        (INT64_MIN, INT64_MAX, "q"),
    ],
)
def test_read_integer_types(min_: int, max_: int, fmt: str) -> None:
    """Tests `read_{B,b,H,h,I,i,Q,q}` and `write_{B,b,H,h,I,i,Q,q}`."""
    num = random.randint(min_, max_)
    converter = CONVERTER[fmt]
    with BytesIOEx(converter.pack(num)) as b:
        read_func: Callable[[], int] = getattr(b, "read_" + fmt)
        assert read_func() == num
        b.seek(0)
        write_func: Callable[[int], int] = getattr(b, "write_" + fmt)
        _ = write_func(num)
        b.seek(0)
        assert read_func() == num


def test_char() -> None:
    """Tests `read_c` and `write_c`."""
    letter = chr(random.randrange(65, 126))
    with BytesIOEx(letter.encode("ascii")) as b:
        b.write_c(letter)
        b.seek(0)
        assert b.read_c() == letter


def test_bool() -> None:
    """Tests `read_bool` and `write_bool`."""
    with BytesIOEx(Bool.pack(True)) as b:
        b.write_bool(False)
        b.seek(0)
        assert not b.read_bool()


def test_double() -> None:
    """Tests `read_d` and `write_d`."""
    random_double = random.random()
    with BytesIOEx(Double.pack(random_double)) as b:
        # A 64-bit floating point integer has about 16 digits of precision
        double = b.read_d()
        assert isinstance(double, float)
        assert math.isclose(double, random_double, rel_tol=1e-16)
        b.write_d(random_double)
        b.seek(1)


def test_float() -> None:
    """Tests `read_f` and `write_f`."""
    random_float = random.random()
    with BytesIOEx(Float.pack(random_float)) as b:
        # A 32-bit floating point integer has about 7 digits of precision
        float_ = b.read_f()
        assert isinstance(float_, float)
        assert math.isclose(float_, random_float, rel_tol=1e-7)
        b.write_f(random_float)
        b.seek(1)


def test_varint() -> None:
    """Tests `read_v` and `write_v`."""
    bignum = random.randint(2**8, 2**32)
    with BytesIOEx() as b:
        b.write_v(bignum)
        b.seek(0)
        assert b.read_v() == bignum


bytesioex/bytesioex.py
import io
from struct import Struct
from typing import Optional

# Converters
Bool = Struct("?")
Char = Struct("c")
SByte = Struct("b")
Byte = Struct("B")
Short = Struct("h")
UShort = Struct("H")
Int = Struct("i")
UInt = Struct("I")
Long = Struct("q")
ULong = Struct("Q")
Float = Struct("f")
Double = Struct("d")


class BytesIOEx(io.BytesIO):
    """A simple wrapper over Python's `io.BytesIO` which provides additional
    methods for reading and writing C data types like `int8`, `uint8`, `bool`,
    `float`, `double` etc."""

    def read_bool(self) -> Optional[bool]:
        """Reads a C99 _Bool."""
        buf = self.read(1)
        return Bool.unpack(buf)[0] if len(buf) == 1 else None

    def read_b(self) -> Optional[int]:
        """Reads a 1-byte signed integer."""
        buf = self.read(1)
        return SByte.unpack(buf)[0] if len(buf) == 1 else None

    def read_B(self) -> Optional[int]:
        """Reads a 1-byte unsigned integer."""
        buf = self.read(1)
        return Byte.unpack(buf)[0] if len(buf) == 1 else None

    def read_c(self) -> Optional[str]:
        """Reads a 1-byte ASCII-character."""
        buf = self.read(1)
        return Char.unpack(buf)[0] if len(buf) == 1 else None

    def read_h(self) -> Optional[int]:
        """Reads a 2-byte signed integer."""
        buf = self.read(2)
        return Short.unpack(buf)[0] if len(buf) == 2 else None

    def read_H(self) -> Optional[int]:
        """Reads a 2-byte unsigned integer."""
        buf = self.read(2)
        return UShort.unpack(buf)[0] if len(buf) == 2 else None

    def read_i(self) -> Optional[int]:
        """Reads a 4-byte signed integer."""
        buf = self.read(4)
        return Int.unpack(buf)[0] if len(buf) == 4 else None

    def read_I(self) -> Optional[int]:
        """Reads a 4-byte unsigned integer."""
        buf = self.read(4)
        return UInt.unpack(buf)[0] if len(buf) == 4 else None

    def read_q(self) -> Optional[int]:
        """Reads an 8-byte signed integer."""
        buf = self.read(8)
        return Long.unpack(buf)[0] if len(buf) == 8 else None

    def read_Q(self) -> Optional[int]:
        """Reads an 8-byte unsigned integer."""
        buf = self.read(8)
        return ULong.unpack(buf)[0] if len(buf) == 8 else None

    def read_f(self) -> Optional[float]:
        """Reads a 4-byte floating-point number."""
        buf = self.read(4)
        return Float.unpack(buf)[0] if len(buf) == 4 else None

    def read_d(self) -> Optional[float]:
        """Reads an 8-byte floating-point number."""
        buf = self.read(8)
        return Double.unpack(buf)[0] if len(buf) == 8 else None

    def read_v(self) -> Optional[int]:
        """Reads a 7-bit encoded integer, the exact size of this type is not
        fixed, hence the name **varint**. If the stream pointer reaches the
        end before being completely able to read, it will return None."""
        b = self.read_B()
        if b is not None:
            data_len = b & 0x7F
            shift = 7
            while (b & 0x80) != 0:
                b = self.read_B()
                if b is None:
                    return None
                data_len |= (b & 0x7F) << shift
                shift += 7
            return data_len
        return None

    def write_bool(self, value: bool) -> int:
        """Converts and writes a `bool` into a C99-style _Bool type,
        where False = 0, True = 1."""
        return self.write(Bool.pack(value))

    def write_b(self, value: int) -> int:
        """Converts and writes a `int` into a signed 8-bit integer."""
        return self.write(SByte.pack(value))

    def write_B(self, value: int) -> int:
        """Converts and writes a `int` into an unsigned 8-bit integer."""
        return self.write(Byte.pack(value))

    def write_c(self, value: str) -> int:
        """Converts and writes a `int` into an unsigned 8-bit integer."""
        return self.write(Char.pack(value))

    def write_h(self, value: int) -> int:
        """Converts and writes a `int` into a signed 16-bit integer."""
        return self.write(Short.pack(value))

    def write_H(self, value: int) -> int:
        """Converts and writes a `int` into an unsigned 16-bit integer."""
        return self.write(UShort.pack(value))

    def write_i(self, value: int) -> int:
        """Converts and writes a `int` into a signed 32-bit integer."""
        return self.write(Int.pack(value))

    def write_I(self, value: int) -> int:
        """Converts and writes a `int` into an unsigned 32-bit integer."""
        return self.write(UInt.pack(value))

    def write_q(self, value: int) -> int:
        """Converts and writes a `int` into a signed 64-bit integer."""
        return self.write(Long.pack(value))

    def write_Q(self, value: int) -> int:
        """Converts and writes a `int` into an unsigned 64-bit integer."""
        return self.write(ULong.pack(value))

    def write_f(self, value: float) -> int:
        """Converts and writes a `float` into a 4-byte floating-point number."""
        return self.write(Float.pack(value))

    def write_d(self, value: float) -> int:
        """Converts and writes a `float` into an 8-byte floating-point number."""
        return self.write(Double.pack(value))

    def write_v(self, buflen: int) -> int:
        """Converts an unsigned integer to a 7-bit encoded integer (varint)
        and writes it.

        Args:
            buflen (int): The length of the buffer to be converted and written.
        """
        ret = bytearray()
        while True:
            towrite = buflen & 0x7F
            buflen >>= 7
            if buflen > 0:
                towrite |= 0x80
            ret.append(towrite)
            if buflen <= 0:
                break
        return self.write(bytes(ret))


bytesioex/__init__.py
"""
BytesIOEx - EXtended `io.BytesIO`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

BytesIOEx is a wrapper over Python's `io.BytesIO` which provides methods
for reading and writing C data types like int8, uint8, bool, double, etc.

### Example usage

    >>> b = BytesIOEx(b"abcd")
    >>> b.read_B()
    97
    >>> b.read_h()
    25442
    >>> b.read()
    b'd'

Full documentation can be found on [ReadTheDocs](https://bytesioex.rtfd.io/).
"""  # noqa

import io
import struct
from typing import Optional

__all__ = [
    "Bool",
    "Char",
    "SByte",
    "Byte",
    "Short",
    "UShort",
    "Int",
    "UInt",
    "Long",
    "ULong",
    "Float",
    "Double",
    "BytesIOEx",
]

# Converters
Bool = struct.Struct("?")
Char = struct.Struct("c")
SByte = struct.Struct("b")
Byte = struct.Struct("B")
Short = struct.Struct("h")
UShort = struct.Struct("H")
Int = struct.Struct("i")
UInt = struct.Struct("I")
Long = struct.Struct("q")
ULong = struct.Struct("Q")
Float = struct.Struct("f")
Double = struct.Struct("d")


# pylint: disable=too-many-public-methods
class BytesIOEx(io.BytesIO):
    """A simple wrapper over Python's `io.BytesIO`.

    Provides methods for reading and writing C data types like `int8`, `uint8`, etc.
    """

    def read_bool(self) -> Optional[bool]:
        """Reads a C99 _Bool."""
        buf = self.read(1)
        return Bool.unpack(buf)[0] if len(buf) == 1 else None

    def read_b(self) -> Optional[int]:
        """Reads a 1-byte signed integer."""
        buf = self.read(1)
        return SByte.unpack(buf)[0] if len(buf) == 1 else None

    def read_B(self) -> Optional[int]:
        """Reads a 1-byte unsigned integer."""
        buf = self.read(1)
        return Byte.unpack(buf)[0] if len(buf) == 1 else None

    def read_c(self) -> Optional[str]:
        """Reads a 1-byte ASCII-character."""
        buf = self.read(1)
        return Char.unpack(buf)[0].decode("ascii") if len(buf) == 1 else None

    def read_h(self) -> Optional[int]:
        """Reads a 2-byte signed integer."""
        buf = self.read(2)
        return Short.unpack(buf)[0] if len(buf) == 2 else None

    def read_H(self) -> Optional[int]:
        """Reads a 2-byte unsigned integer."""
        buf = self.read(2)
        return UShort.unpack(buf)[0] if len(buf) == 2 else None

    def read_i(self) -> Optional[int]:
        """Reads a 4-byte signed integer."""
        buf = self.read(4)
        return Int.unpack(buf)[0] if len(buf) == 4 else None

    def read_I(self) -> Optional[int]:
        """Reads a 4-byte unsigned integer."""
        buf = self.read(4)
        return UInt.unpack(buf)[0] if len(buf) == 4 else None

    def read_q(self) -> Optional[int]:
        """Reads an 8-byte signed integer."""
        buf = self.read(8)
        return Long.unpack(buf)[0] if len(buf) == 8 else None

    def read_Q(self) -> Optional[int]:
        """Reads an 8-byte unsigned integer."""
        buf = self.read(8)
        return ULong.unpack(buf)[0] if len(buf) == 8 else None

    def read_f(self) -> Optional[float]:
        """Reads a 4-byte floating-point number."""
        buf = self.read(4)
        return Float.unpack(buf)[0] if len(buf) == 4 else None

    def read_d(self) -> Optional[float]:
        """Reads an 8-byte floating-point number."""
        buf = self.read(8)
        return Double.unpack(buf)[0] if len(buf) == 8 else None

    def read_v(self) -> Optional[int]:
        """Reads a 7-bit encoded integer.

        The exact underlying size of this type is not fixed, hence the name
        **varint**. If the stream pointer reaches the end before being
        completely able to read, returns `None`.
        """
        b = self.read_B()
        if b is not None:
            data_len = b & 0x7F
            shift = 7
            while (b & 0x80) != 0:
                b = self.read_B()
                if b is None:
                    return None
                data_len |= (b & 0x7F) << shift
                shift += 7
            return data_len
        return None

    def write_bool(self, value: bool) -> int:
        """Converts and writes a `bool` into a C99-style _Bool type.

        False = 0, True = 1.
        """
        return self.write(Bool.pack(value))

    def write_b(self, value: int) -> int:
        """Converts and writes a `int` into a signed 8-bit integer."""
        return self.write(SByte.pack(value))

    def write_B(self, value: int) -> int:
        """Converts and writes a `int` into an unsigned 8-bit integer."""
        return self.write(Byte.pack(value))

    def write_c(self, value: str) -> int:
        """Converts and writes a `str` into a char."""
        return self.write(Char.pack(value.encode("ascii")))

    def write_h(self, value: int) -> int:
        """Converts and writes a `int` into a signed 16-bit integer."""
        return self.write(Short.pack(value))

    def write_H(self, value: int) -> int:
        """Converts and writes a `int` into an unsigned 16-bit integer."""
        return self.write(UShort.pack(value))

    def write_i(self, value: int) -> int:
        """Converts and writes a `int` into a signed 32-bit integer."""
        return self.write(Int.pack(value))

    def write_I(self, value: int) -> int:
        """Converts and writes a `int` into an unsigned 32-bit integer."""
        return self.write(UInt.pack(value))

    def write_q(self, value: int) -> int:
        """Converts and writes a `int` into a signed 64-bit integer."""
        return self.write(Long.pack(value))

    def write_Q(self, value: int) -> int:
        """Converts and writes a `int` into an unsigned 64-bit integer."""
        return self.write(ULong.pack(value))

    def write_f(self, value: float) -> int:
        """Converts and writes a `float` into a 4-byte floating-point number."""
        return self.write(Float.pack(value))

    def write_d(self, value: float) -> int:
        """Converts and writes a `float` into an 8-byte floating-point number."""
        return self.write(Double.pack(value))

    def write_v(self, buflen: int) -> int:
        """Converts an unsigned integer to a 7-bit encoded integer (varint).

        Args:
            buflen (int): Length of the buffer.
        """
        ret = bytearray()
        while True:
            towrite = buflen & 0x7F
            buflen >>= 7
            if buflen > 0:
                towrite |= 0x80
            ret.append(towrite)
            if buflen <= 0:
                break
        return self.write(bytes(ret))
